docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+Payload values that are smaller than the payload space are then stored aligned to the opposite end of the value to the discriminator.
+
+*Rationale:* This case is reached if we are packing a union type inside the significand bits of a signalling NaN and guarantees that we can load the payload in a single aligned load and need to zero bits at only one end to extract the payload.
+
+In other cases, the discriminator bits are stored in the low bits (as defined by the platform byte order) of the object and the payload in the high bits.
+
+*Rationale:* The discriminator is easy to mask off with a right shift and extract with an xor with an immediate operand for comparisons.
+
+The types identified in a discriminator are sorted lexically and discriminator values are assigned in order.
+Singletons are sorted before all other types.
+
+*Rationale:* The representation of `A | B` and `B | A` should be the same.
+It is useful to sort singletons before pointers so that the common case of `T* | NotAThing` is able to use null-pointer checks and branches, which are typically optimised aggressively in modern microarchitectures.
+
+{*Note:* I'm not sure lexical sorting will work because types imported from the same package by two other packages will have different names and so we may need some other stable ordering for types.}
+
+*Open question*: As described, this requires a pointer dereference do disambiguate pointer types.
+It would be possible to use any space in the low three bits that is not used to disambiguate non-pointer types to differentiate a small number of pointee types.
+This would save a load in pattern matching and might be useful, though given that the main reason for pattern matching is to then call a method on the concrete type, it's not clear that this is actually a win.
+Note that, in the CHERI case, the low bits are still available and so it may be useful to use them to differentiate between pointee types.
+
+These rules are somewhat complex in the abstract.
+The following sections give some concrete examples of how they are applied.
+
+### Example `U64 | Float64`
+
+The size of `U64` is 64 bits and it requires one discriminator state.
+The size of `Float64` is 64 bits and it requires one discriminator state.
+
+We first try the special-case rule for the case that the largest value is a floating-point value.
+The remaining type (`U64`) requires 64 bits for payload and 0 bits for discriminator, and so does not fit into the significand part of a `Float64` signalling NaN.
+The first special case is therefore not used
+The second special case is not used because neither type is a pointer.
+
+We are now left requiring one bit of discriminator and 64 bits of payload.
+The payload size is a power of two, and so the address of the union type is the address of the start of a 64-bit payload.
+The 1-bit discriminator is stored afterwards, with a value of 0 representing a `Float64` and a value of `1` representing a `U64`.
```

Comment:
I believe this means each case will have a special encoding, from 0 to N-1 for N types, which is the point I was trying to make above. 

This "works" if we keep a map of every union in the compiler (or reconstruct the representation every time) and know how to check types and extract the payload. But when looking at the code generated, it'd be hard to know what's the shift and mask for each type, many of which could be merged with some other instruction by the optimising compiler.

I wish we could have a single representation, using the 4-bits in the high-bit-gap of the FP64 representation for all types that fit the payload (`{I,U}x{8,16,32}` + `{F16, F32}` + `ptr`) and do pointer and class separation in the same way we'll do `{I,U}x{Size,64,128}` and `F128`, which is, with an additional descriptor.

This means, I'd prefer to have a separate encoding for boxed primitives and boxed complex types, including singleton types.

I'm assuming that developers understand that usually `for (int in list) -> sum+=int` is faster than `for (obj in list) -> sum+=obj.field`, and that the additional run-time requirements of indirectly calling methods or accessing fields as well as the memory requirements for dynamic classes (perhaps excluding singletons) won't have a huge impact.

Reply:
I see, each variation of the union with 1 or 2 types would require a change in representation, but with 2 types you (may) have to deconstruct first, then reconstruct. As the number of types in the union grows, and the number of different usages grow, this can become a bottleneck.

What I'm worried about is also the logic to map N to M conversions where both N and M are unbound. Having to deconstruct to the naive representation and back into another packed form would make the compiler a lot simpler, but would also be worst for run-time performance. 

Unless we come up with a generic algorithm (which I'm not sure yet we can), we'll end up catering for a few particular pain points and leaving the rest as naive conversions. That works, I guess, but it'll be a source of constant pain.